util/codec/codec.go

```diff
@@ -870,27 +870,31 @@ func DecodeAsDateTime(b []byte, tp byte, loc *time.Location) (remain []byte, d t
 	}
 	flag := b[0]
 	b = b[1:]
+	var v uint64
 	switch flag {
 	case uintFlag:
-		var v uint64
 		b, v, err = DecodeUint(b)
-		if err != nil {
-			return b, d, err
-		}
-		t := types.NewTime(types.ZeroCoreTime, tp, 0)
-		err = t.FromPackedUint(v)
-		if err == nil {
-			if tp == mysql.TypeTimestamp && !t.IsZero() && loc != nil {
-				err = t.ConvertTimeZone(time.UTC, loc)
-				if err != nil {
-					return b, d, err
-				}
-			}
-			d.SetMysqlTime(t)
-		}
+	case uvarintFlag:
+		// Datetime can be encoded as Uvarint
+		b, v, err = DecodeUvarint(b)
+
 	default:
 		return b, d, errors.Errorf("invalid encoded key flag %v", flag)
 	}
+	if err != nil {
+		return b, d, err
+	}
+	t := types.NewTime(types.ZeroCoreTime, tp, 0)
+	err = t.FromPackedUint(v)
+	if err == nil {
+		if tp == mysql.TypeTimestamp && !t.IsZero() && loc != nil {
+			err = t.ConvertTimeZone(time.UTC, loc)
+			if err != nil {
+				return b, d, err
+			}
+		}
+		d.SetMysqlTime(t)
+	}
 	if err != nil {
 		return b, d, errors.Trace(err)
 	}
```

Comment:
```
if err == nil {
  ...
} else {
  return b, d, errors.Trace(err)
}
```

Reply:
Thanks, I updated the code, but used something like the below instead:
```
if err != nil {
  return b, d, errors.Trace(err)
}
... // continue with the logic
```